ambry-clustermap/src/main/java/com.github.ambry.clustermap/AmbryDisk.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright 2017 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.clustermap;
+
+import com.github.ambry.config.ClusterMapConfig;
+import com.github.ambry.utils.Utils;
+import java.io.File;
+
+
+/**
+ * {@link DiskId} implementation to use within dynamic cluster managers.
+ */
+class AmbryDisk implements DiskId, Resource {
+  private final AmbryDataNode datanode;
+  private final String mountPath;
+  private final long rawCapacityBytes;
+  private final ResourceStatePolicy resourceStatePolicy;
+
+  /**
+   * Instantiate an AmbryDisk object.
+   * @param clusterMapConfig the {@link ClusterMapConfig} to use.
+   * @param datanode the {@link AmbryDataNode} on which this disk resides.
+   * @param mountPath the mount path at which this disk is mounted on the datanode.
+   * @param state the initial {@link HardwareState} of this disk.
+   * @param rawCapacityBytes the capacity in bytes.
+   * @throws Exception if there is an exception in instantiating the {@link ResourceStatePolicy}
+   */
+  AmbryDisk(ClusterMapConfig clusterMapConfig, AmbryDataNode datanode, String mountPath, HardwareState state,
+      long rawCapacityBytes) throws Exception {
+    this.datanode = datanode;
+    this.mountPath = mountPath;
+    this.rawCapacityBytes = rawCapacityBytes;
+    ResourceStatePolicyFactory resourceStatePolicyFactory =
```

Comment:
Actually this can't be passed in as is (contrary to what I led you to believe) as `AmbryDisk` and `AmbryDatanode` implement the `Resource` interface and the factory needs that resource.